enhancements/installer/openstack-ipi-failure-domains.md

```diff
@@ -0,0 +1,260 @@
+
+## Proposal
+
+The Installer allow users to provide a list of Failure Domains which contain information about networking and storage.
+The cluster will be deployed in these domains on day 1 via IPI.
+The Installer will validate the information given by the user and return an error if a resource doesn't exist or
+is incorrectly used.
+
+The cluster will be deployed in the different Failure Domains, following the Round Robin process. The first master will be deployed in the first domain, and once we have used all domains, we come back to the first one, etc.
+
+When using Failure Domains, the users will have to do it for both the masters and the workers.
+
+The infrastructure resources owned by the cluster continue to be clearly identifiable and distinct from other resources.
+
+Destroying a cluster must make sure that no resources are deleted that didn't belong to the cluster.  Leaking resources is preferred over any possibility of deleting non-cluster resources.
+
+### Workflow Description
+
+- The OpenStack administrators will deploy OpenStack with multiple Availability Zones, so Nova and Cinder can respectfully deploy servers and volumes in the Failure Domain.
+- The OpenStack administrators will create at least one Routed Provider Network, and then multiple subnets, where each zone has a least one subnet.
+- The OpenShift administrators will identify the Domain Failures: their availability zones, subnets, etc. They'll decide which ones will be used for the masters, and the ones for the workers (domains can be shared).
+- The OpenShift administrators will provide the right configuration in the `install-config.yaml` file and then deploy the cluster (detailed later in this document).
+- (Out of scope for this enhancement, see "Non-Goals") The OpenShift administrators will make sure that the network infrastructure has a networking route to reach the OCP Control Plane VIPs.
+
+### API Extensions
+
+Not applicable.
+
+
+### Implementation Details/Notes/Constraints
+
+#### Types
+
+```golang
+// OpenStackPlatformFailureDomainSpec holds the zone failure domain and
+// the Neutron subnet of that failure domain.
+type OpenStackPlatformFailureDomainSpec struct {
+  // name defines the name of the OpenStackPlatformFailureDomainSpec
+  Name        string `json:"name"`
+  ComputeZone string `json:"computeZone"`
+  StorageZone string `json:"storageZone"`
+  Subnet      string `json:"subnet"`
+}
+```
+
+#### Resource provided to the installer
+
+The OpenShift administrators provide a list of `failureDomains` and their information:
+
+```yaml
+platform:
+  openstack:
+    ...
+    failureDomains:
+    - name: rack1
+      computeZone: az1
+      storageZone: cinder-az1
+      subnet: d7ffd3d8-87e1-481f-a818-ff2f17787d40
+    - name: rack2
+      computeZone: az2
+      storageZone: cinder-az2
+      subnet: f584ce8c-bb54-422b-8e11-6a530be4c23e
```

Comment:
This design "defines" the `failureDomains` in `platform` and then includes references in the machine pools. 

Did you consider just including the definition in the machine pools and omitting the `failureDomains` from the platform? Only including failureDomains in the machine pool seems simpler to me.

Reply:
This would mean duplicating `failureDomains` between workers and masters for example, if they share the same attributes (AZ, subnet, etc).

Reply:
If the failureDomains were the same you could use the defaultMachinePool instead of both compute and controlPlane but I suppose this would not avoid duplication in every scenario.

Reply:
My original comment was a typo. I intended `defaultMachinePlatform`--not `defaultMachinePool` which is not a thing.

Reply:
I know @mandre has some local change he wanted to push but please look at these comments before ^
I think using `defaultMachinePlatform` makes sense, and we need to come up with an easy naming to differentiate the list of domains, and their definitions (with attributes, like az, etc)

Reply:
@patrickdillon any thoughts on this? ^

Reply:
In case we go for defining every `failureDomain` right in the machine-pool that uses it, then we can document how to use YAML anchors to avoid repetition.

Example:
```YAML
apiVersion: v1
compute:
- architecture: amd64
  hyperthreading: Enabled
  name: worker
  platform:
    openstack:
      type: compute.flavor
      failureDomains:
      - &rack1
        computeZone: az1
        storageZone: cinder-az1
        subnet: aaaaaaaa-87e1-481f-a818-ff2f17787d40
      - &rack2
        computeZone: az2
        storageZone: cinder-az2
        subnet: bbbbbbbb-87e1-481f-a818-ff2f17787d40
      - &rack3
        computeZone: az3
        storageZone: cinder-az3
        subnet: cccccccc-87e1-481f-a818-ff2f17787d40
  replicas: 3
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform:
    openstack:
      type: controlplane.flavor
      failureDomains:
      - *rack1
      - <<: *rack3
        storageZone: cinder-az3b
  replicas: 3
```

compiles to:

```YAML
apiVersion: v1
compute:
  - architecture: amd64
    hyperthreading: Enabled
    name: worker
    platform:
      openstack:
        type: compute.flavor
        failureDomains:
          - computeZone: az1
            storageZone: cinder-az1
            subnet: aaaaaaaa-87e1-481f-a818-ff2f17787d40
          - computeZone: az2
            storageZone: cinder-az2
            subnet: bbbbbbbb-87e1-481f-a818-ff2f17787d40
          - computeZone: az3
            storageZone: cinder-az3
            subnet: cccccccc-87e1-481f-a818-ff2f17787d40
    replicas: 3
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform:
    openstack:
      type: controlplane.flavor
      failureDomains:
        - computeZone: az1
          storageZone: cinder-az1
          subnet: aaaaaaaa-87e1-481f-a818-ff2f17787d40
        - computeZone: az3
          storageZone: cinder-az3b
          subnet: cccccccc-87e1-481f-a818-ff2f17787d40
  replicas: 3
```

Reply:
I hope we can avoid that by defining them in `Platform` so we have no duplication...

Reply:
Why avoiding that? I would love delegating the whole deduplication matter to the language itself.

Reply:
From a user perspective, I think it's better to define this setting at the Platform level and avoid YAML anchors where not everyone is familiar with. I don't see any overhead using Platform, but maybe I missed something.